File: src/leetCode/java/Solution253.java
package leetCode.java;

import java.util.*;

//Intervals O(nlog) O(n), sort需要O(nlogn) O(n)
class Solution253 {//lintcode919
    public int minMeetingRooms(List<Interval> intervals) {
        //將start與end分別拿出排序
        int[] start = new int[intervals.size()];
        int[] end = new int[intervals.size()];
        for(int i=0; i<intervals.size(); i++){
            start[i] = intervals.get(i).start;
            end[i] = intervals.get(i).end;
        }
        Arrays.sort(start);
        Arrays.sort(end);

        //模擬
        int s = 0;
        int e = 0;
        int room = 0;
        int res = 0;
        while(s < start.length){
            if(start[s] < end[e]){
                s++;
                room++;
                res = Math.max(res, room);
            }else{
                e++;
                room--;
            }
        }
        return res;
    }
}
/*
0-------------------30
	5---10
		10------20

*/

File: src/leetCode/c++/Solution253.cpp
//Intervals O(nlog) O(n), sort需要O(nlogn) O(n)
class Solution253 {//lintcode919
public:
    int minMeetingRooms(vector<Interval> &intervals) {
        vector<int> start;
        vector<int> end;
        for(Interval interval : intervals){
            start.push_back(interval.start);
            end.push_back(interval.end);
        }

        sort(start.begin(), start.end());
        sort(end.begin(), end.end());

        int s = 0, e = 0, cnt = 0, res = 0;
        while(s < start.size()){
            if(start[s] < end[e]){
                ++cnt;
                ++s;
                res = max(res, cnt);
            }else{
                --cnt;
                ++e;
            }
        }
        return res;
    }
};
/* intervals = [(0,30),(5,10),(15,20)]
0   5   10  15  20      30
--------------------------
    ------  ------

*/